### PR TITLE
Handle dynamic IAM secrets

### PR DESCRIPTION
**Pull request by 7fELF**
Allows the Exoscale IAM secret to be loaded from a file, and automatically reloaded when this file is modified.

The configuration file is a JSON file that follows the format of the secret created by the [exoscale IAM vault plugin](https://github.com/exoscale/vault-plugin-secrets-exoscale)

Templating that file from vault ends up being as easy as:
```
{{- with secret "exoscale/apikey/my-ccm-role" -}}
{{ .Data | toJSON }}
{{- end }}
```

All call to egoscale are now guarded by a Mutex preventing its use while the IAM credential are being updated.

### CODE REVIEW

File: exoscale/client.go [hunk 1]
```diff
@@ -1,29 +1,162 @@
...
+	err = json.NewDecoder(f).Decode(&credentials)
+	if err != nil {
+		fatalf("failed to decode credentials file %q: %s", e.credentialsFile, err)
+	}
+	f.Close()
+
+	e.client = egoscale.NewClient(e.endpoint, credentials.APIKey, credentials.APISecret)
+	e.Unlock()
+
+	infof("exoscale IAM credentials refreshed, now using %s (%s)", credentials.Name, credentials.APIKey)
```
Comment (falzm) [inline exoscale/client.go:118]:
```suggestion
	infof("Exoscale API credentials refreshed, now using %s (%s)", credentials.Name, credentials.APIKey)
```

File: exoscale/client.go [hunk 2]
```diff
@@ -1,29 +1,162 @@
...
+				(event.Op&fsnotify.Write == fsnotify.Write || event.Op&fsnotify.Create == fsnotify.Create) {
+				infof("refreshing IAM credential from file %s", e.credentialsFile)
+				e.refreshCredentials()
+			}
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				errorf("credentials file watcher error channel closed")
+				return
+			}
+			errorf("error while watching credentials file %q, %q", e.credentialsFile, err)
```
Comment (falzm) [inline exoscale/client.go:90]:
```suggestion
			errorf("error while watching credentials file %q: %v", e.credentialsFile, err)
```

File: exoscale/client.go [hunk 3]
```diff
@@ -1,29 +1,162 @@
...
 	}
 
-	egoscale.UserAgent = fmt.Sprintf("Exoscale-K8s-Cloud-Controller/%s %s", versionString, egoscale.UserAgent)
+	return exoclient, nil
+}
+
+func (e *exoscaleClient) watchCredentialsFile(stop <-chan struct{}) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		fatalf("failed to watch credentials file %q, %q", e.credentialsFile, err)
```
Comment (falzm) [inline exoscale/client.go:63]:
```suggestion
		fatalf("failed to watch credentials file %q: %v", e.credentialsFile, err)
```

File: exoscale/client.go [hunk 4]
```diff
@@ -1,29 +1,162 @@
...
+
+func (e *exoscaleClient) UpdateNetworkLoadBalancer(ctx context.Context, zone string,
+	nlbUpdate *egoscale.NetworkLoadBalancer) (*egoscale.NetworkLoadBalancer, error) {
+	e.RLock()
+	defer e.RUnlock()
+
+	return e.client.UpdateNetworkLoadBalancer(ctx, zone, nlbUpdate)
+}
+
+func (e *exoscaleClient) GetVirtualMachine(ctx context.Context, uuid *egoscale.UUID) (*egoscale.VirtualMachine, error) {
```
Comment (falzm) [inline exoscale/client.go:151]: I know that the underlying type is still `egoscale.VirtualMachine` for historical reasons (which will change in API v2), however we do our best to normalize the name everywhere as *[Compute] instance*:

```suggestion
func (e *exoscaleClient) GetInstance(ctx context.Context, uuid *egoscale.UUID) (*egoscale.VirtualMachine, error) {
```

File: exoscale/client.go [hunk 5]
```diff
@@ -1,29 +1,162 @@
...
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				errorf("credentials file watcher event channel closed")
+				return
+			}
+
+			if event.Name == e.credentialsFile &&
+				(event.Op&fsnotify.Write == fsnotify.Write || event.Op&fsnotify.Create == fsnotify.Create) {
+				infof("refreshing IAM credential from file %s", e.credentialsFile)
```
Comment (falzm) [inline exoscale/client.go:82]:
```suggestion
				infof("refreshing API credentials from file %q", e.credentialsFile)
```

File: exoscale/client.go [hunk 6]
```diff
@@ -1,29 +1,162 @@
...
-		envEndpoint = defaultComputeEndpoint
+	if envEndpoint != "" {
+		exoclient.endpoint = envEndpoint
 	}
 
-	if envKey == "" || envSecret == "" {
+	egoscale.UserAgent = fmt.Sprintf("Exoscale-K8s-Cloud-Controller/%s %s", versionString, egoscale.UserAgent)
+
+	if envKey != "" && envSecret != "" {
+		infof("reading exoscale IAM credentials from environment")
```
Comment (falzm) [inline exoscale/client.go:46]:
```suggestion
		infof("reading Exoscale API credentials from environment")
```

File: exoscale/client.go [hunk 7]
```diff
@@ -1,29 +1,162 @@
...
+		exoclient.client = egoscale.NewClient(exoclient.endpoint, envKey, envSecret)
+	} else if envCredentialsFile != "" {
+		exoclient.credentialsFile = envCredentialsFile
+		infof("reading exoscale IAM credentials from file %q", exoclient.credentialsFile)
+		exoclient.refreshCredentials()
+		go exoclient.watchCredentialsFile(stop)
+	} else {
 		return nil, errors.New("incomplete or missing for API credentials")
```
Comment (falzm) [inline exoscale/client.go:54]: Not changed in your PR, but it's a good opportunity to fix this typo :)

```suggestion
		return nil, errors.New("incomplete or missing Exoscale API credentials")
```

File: exoscale/client.go [hunk 8]
```diff
@@ -1,29 +1,162 @@
...
 
-	if envKey == "" || envSecret == "" {
+	egoscale.UserAgent = fmt.Sprintf("Exoscale-K8s-Cloud-Controller/%s %s", versionString, egoscale.UserAgent)
+
+	if envKey != "" && envSecret != "" {
+		infof("reading exoscale IAM credentials from environment")
+		exoclient.client = egoscale.NewClient(exoclient.endpoint, envKey, envSecret)
+	} else if envCredentialsFile != "" {
+		exoclient.credentialsFile = envCredentialsFile
+		infof("reading exoscale IAM credentials from file %q", exoclient.credentialsFile)
```
Comment (falzm) [inline exoscale/client.go:50]:
```suggestion
		infof("reading Exoscale API credentials from file %q", exoclient.credentialsFile)
```

**Review by falzm — CHANGES_REQUESTED** [review]
Do you plan on testing this new scenario in the integration tests suite?

File: exoscale/common.go [hunk 9]
```diff
@@ -25,12 +25,12 @@ func (c *cloudProvider) computeInstanceByProviderID(ctx context.Context, provide
 		return nil, err
 	}
 
-	r, err := c.client.GetWithContext(ctx, egoscale.VirtualMachine{ID: uuid})
+	r, err := c.client.GetVirtualMachine(ctx, uuid)
```
Comment (7fELF) [inline exoscale/common.go:28]:
```suggestion
	r, err := c.client.GetInstance(ctx, uuid)
```

File: exoscale/loadbalancer.go [hunk 10]
```diff
@@ -97,9 +97,9 @@ func (l *loadBalancer) EnsureLoadBalancer(ctx context.Context, _ string, service
 
 		instancePoolID := ""
 		for _, node := range nodes {
-			instance, err := l.fetchComputeInstanceFromNode(ctx, node)
+			instance, err := l.p.client.GetVirtualMachine(ctx, egoscale.MustParseUUID(node.Status.NodeInfo.SystemUUID))
```
Comment (7fELF) [inline exoscale/loadbalancer.go:100]:
```suggestion
			instance, err := l.p.client.GetInstance(ctx, egoscale.MustParseUUID(node.Status.NodeInfo.SystemUUID))
```

**Review by exo-cedric — APPROVED** [review]
lgtm, afaict. thx

**Review by sgmac — APPROVED** [review]

**Review by pierre-emmanuelJ — APPROVED** [review]

**Review by falzm — APPROVED** [review]
LGTM 👍

**Pull request merged**